compiler/rustc_index/src/vec.rs

```diff
@@ -817,34 +815,39 @@ impl<'a, I: Idx, T> IntoIterator for &'a mut IndexVec<I, T> {
     }
 }
 
-pub struct IntoIdx<I: Idx> {
+pub struct Enumerated<I: Idx, It: Iterator> {
+    iter: iter::Enumerate<It>,
     _marker: PhantomData<fn(&I)>,
 }
-impl<I: Idx, T> FnOnce<((usize, T),)> for IntoIdx<I> {
-    type Output = (I, T);
 
-    extern "rust-call" fn call_once(self, ((n, t),): ((usize, T),)) -> Self::Output {
-        (I::new(n), t)
+impl<I: Idx, It: Iterator> Iterator for Enumerated<I, It> {
```

Comment:
Does this implementation allow the same kind of optimizations by specializations as the earlier one?
Should we rather make `IntoIdx` a function and use RPIT?

Reply:
> Does this implementation allow the same kind of optimizations by specializations as the earlier one?

Probably not.

> Should we rather make IntoIdx a function and use RPIT?

I will try this.

Reply:
Done